### PR TITLE
Updated ensime bindings to better fit with spacemacs

### PR DESCRIPTION
**Pull request by bjarkevad**
I've updated the bindings for ensime so they fit a bit better with spacemacs, but rather than following spacemacs convetions I've pretty much just translated ensime's default bindings from `C-c C-{X} something` to `spc m {X} something`.
I'm not sure if this is the best idea, but it makes switching from normal emacs + ensime to spacemacs + ensime easier.

### CODE REVIEW

**Comment by chrisbarrett** [comment]
I like this. +1

File: contrib/lang/scala/packages.el [hunk 1]
```diff
@@ -47,6 +47,66 @@ which require an initialization must be listed explicitly in the list.")
...
         (kbd "N") 'backward-button)
 
+      (evil-leader/set-key-for-mode 'scala-mode
+        "mri"     'ensime-refactor-inline-local
+        "mrl"     'ensime-refactor-extract-local
+        "mrm"     'ensime-refactor-extract-method
+        "mro"     'ensime-refactor-organize-imports
+        "mrr"     'ensime-refactor-rename
+        "mrt"     'ensime-import-type-at-point
+
```
Comment (syl20bnr) [inline contrib/lang/scala/packages.el]: It is not in the conventions for now but the refactoring is planned to be on `SPC m R` instead of `SPC m r`.

File: contrib/lang/scala/packages.el [hunk 2]
```diff
@@ -47,6 +47,66 @@ which require an initialization must be listed explicitly in the list.")
...
+        "mbS"     'ensime-stacktrace-switch
+        "mbT"     'ensime-sbt-do-test
+        "mbc"     'ensime-sbt-do-compile
+        "mbn"     'ensime-sbt-do-clean
+        "mbo"     'ensime-sbt-do-test-only
+        "mbp"     'ensime-sbt-do-package
+        "mbr"     'ensime-sbt-do-run
+        "mbs"     'ensime-sbt-switch
+        "mbt"     'ensime-sbt-do-test-quick
+
```
Comment (syl20bnr) [inline contrib/lang/scala/packages.el]: I like `mb` for sbt commands!

File: contrib/lang/scala/packages.el [hunk 3]
```diff
@@ -47,6 +47,66 @@ which require an initialization must be listed explicitly in the list.")
...
+        "mdo"     'ensime-db-step-out
+        "mdq"     'ensime-db-quit
+        "mdr"     'ensime-db-run
+        "mds"     'ensime-db-step
+        "mdt"     'ensime-db-backtrace
+        "mdu"     'ensime-db-clear-break
+
+        "mti"     'ensime-goto-impl
+        "mtt"     'ensime-goto-test
+
```
Comment (syl20bnr) [inline contrib/lang/scala/packages.el]: `mt` is reserved for tests.
At first sight these commands should go behind `mg`

**Comment by syl20bnr** [comment]
I will wait a bit before merging it because I have to read the doc of each command and maybe play a bit with them in order to make them match with what we have in the other layers.

Each new added tool is a great opportunity to discover new commands so this PR has a lot of value. Thank you !

**Comment by bjarkevad** [comment]
Thanks for the feedback!
As a sidenote, when I was reading the ensime docs I noticed that a lot of times only the keybinding was refered to, so it was hard to translate ensime bindings to spacemacs-convention bindings without making ensime docs hard to follow.

File: contrib/lang/scala/packages.el (hunk 3 above)
Comment (bjarkevad) [inline contrib/lang/scala/packages.el]: So `ensime-goto-{test,impl}` is used to jump between test and implementation files, so it's not a mismatch with spacemacs conventions if we want to keep them this way.

File: contrib/lang/scala/packages.el [hunk 4]
```diff
@@ -47,6 +47,78 @@ which require an initialization must be listed explicitly in the list.")
         (kbd "n") 'forward-button
         (kbd "N") 'backward-button)
 
+      (defun ensime-gen-and-reload()
```
Comment (bjarkevad) [inline contrib/lang/scala/packages.el]: I've added a new function which generates an .ensime and reloads ensime, which is useful when the project changes. Sometimes it's also necessary to `rm -r .ensime_cache` in the project root to force ensime to notice changes to e.g. dependencies.

File: contrib/lang/scala/packages.el [hunk 5]
```diff
@@ -47,6 +47,78 @@ which require an initialization must be listed explicitly in the list.")
...
+        (interactive)
+        (progn
+          (sbt-command "gen-ensime")
+          (ensime-shutdown)
+          (ensime))
+        )
+      
+      (evil-leader/set-key-for-mode 'scala-mode
+        "mg"     'ensime-edit-definition
```
Comment (bjarkevad) [inline contrib/lang/scala/packages.el]: This is a temporary binding, but the function should be included in the bindings

File: contrib/lang/scala/packages.el [hunk 6]
```diff
@@ -47,6 +47,78 @@ which require an initialization must be listed explicitly in the list.")
...
+
+        "m."      'ensime-gen-and-reload
+        "m,"      'ensime
```
Comment (bjarkevad) [inline contrib/lang/scala/packages.el]: Using `, ,` and `, .` for ensime-server commands. Could probably be moved to some more mnemonic binding.

File: contrib/lang/scala/packages.el (hunk 5 above)
Comment (bjarkevad) [inline contrib/lang/scala/packages.el]: This is a temporary binding, but the function should be included in the bindings

**Comment by syl20bnr** [comment]
Excellent PR ! :+1:  :heart: 
I adapted it to the conventions in this commit: 4be4abc12b51635199d5ff35c5d4376ca60f6def

Cherry-picked into `develop`.

**Comment by syl20bnr** [comment]
@bjarkevad, forgot to mention that I decided to put refactoring bindings under `mr` instead of `mR` since it does not make sense to use `R` which is harder to press.
I originally planned to put them under `mR` because ruby already uses `mr` for RoR bindings, but it makes little sense to shift all refactoring bindings on `mR` just for ruby.

CC @bru We have 2 solutions:
1) move RoR bindings to another key that is easy to press with m (could be `mn`, `mm` etc...) but not mnemonic
2) make an exception with ruby and put refactoring on `mR`

I would go with 1) to keep consistency with `mr` everywhere.

**Comment by bru** [comment]
Considered the fact that currently the ruby layer has 1 refactor command vs
a few dozen Rails commands, I would yield precedence to the latter.
That said, 'mm' could be a better prefix given: 1. the (very high)
frequency we need code navigation commands and 2. the fact the ',' shortcut
still doesn't work in all modes (so I end up always using the long form)

**Comment by syl20bnr** [comment]
`,` should work whenever there is `SPC m` for the current mode, if not it is a big bug :-(
Do you have examples where there are available `SPC m` bindings available and `,` does not work ?

**Comment by bru** [comment]
Essentially any view in a rails project.
I thought that was happening because a views's major mode is HTML/web, but
the long form (SPC m ...) still works, so I'm a bit puzzled. Happy to help
you debugging this, of course.

**Comment by syl20bnr** [comment]
I never thought about this... this is a big limitation, `SPC m` is major-mode dependent. This is really bad :-(

Extending `evil-leader` to minor-modes can be useful then but we have to be extra-careful about using it.
What rule do you have to enable rails command as a minor mode in the views ?
